src/main/java/org/eclipse/yasson/internal/model/PropertyModel.java

```diff
@@ -117,6 +131,19 @@ public PropertyModel(ClassModel classModel, Property property, JsonbContext json
         return null;
     }
 
+    private Type resolveType() {
+        if (getterType instanceof ParameterizedType
+                && !propertyType.equals(getterType)
+                && Optional.class.isAssignableFrom(ReflectionUtils.getRawType(getterType))) {
+            if (!propertyType.equals(((ParameterizedType) getterType).getActualTypeArguments()[0])) {
+                throw new JsonbException(Messages.getMessage(MessageKeys.GETTER_RETURN_TYPE_DOESNT_MATCH_FIELD_TYPE));
```

Comment:
Should be don in class parsing phase, not in runtime when properties are accessed constantly.